### ISSUE TITLE
Possible fundamental bug in your code

### ISSUE THREAD

**Issue by tuhinjubcse** [pinkeshbadjatiya/twitter-hatespeech#6]
Hi
Pinkesh 

Your cross validation strategy is correct but you have initialized your model in lstm.py outside the CV  loop and in train_LSTM method that same model is getting trained for every fold which means your model parameters are exposed to the entire dataset ,eventually meaning that your test set consists of samples trained on in previous folds . This is faulty and hence it explains why the F1 numbers increase gradually as the folds increase . Please let me know if my understanding is correct and if so how are you trying to fix it because your numbers will vary accordingly .

**Comment by pinkeshbadjatiya** [comment]
Hello,

Thanks for going through the code. I believe we are shuffling the weights as an alternate way to create a new instance of the model. That should be enough I believe.

**Comment by tuhinjubcse** [comment]
if INITIALIZE_WEIGHTS_WITH == "glove":
            model.layers[0].set_weights([weights])
        elif INITIALIZE_WEIGHTS_WITH == "random":
            shuffle_weights(model)
        else:
            print "ERROR!"
            return

Does your hypothesis hold even if we use glove ?

**Comment by pinkeshbadjatiya** [comment]
I missed the GLOVE initialization part. As of now, I agree that is a valid bug in the code as it only resets the 1st layer. I am not sure if this is the exact version we used in the paper, I'll have to run the code to see if things change.

It is possible that this is just a bug in the code shared on GitHub. Our final version before sharing on Github involved the pipeline of the models, which I then stripped to make it simple. It might be possible that I stripped that line during the process as the code was a bit messy (still it is). It is still just a possibility.

**Comment by tuhinjubcse** [comment]
You should share the exact code as it is helpful for researchers who are going ahead and working on the same problem to compare against your methods . Hope you will look into this and change the code and report the true numbers   .  I will try to reproduce your approach with random embeddings and let you know of any discrepancy .

**Comment by pinkeshbadjatiya** [comment]
I am still pretty sure this just happened while sharing the code as the code was a bit messy making it difficult to restructure. I added the API calls before sharing on GitHub to make it easier for others with no knowledge of code to execute it using simple args. We specifically never used the args format during our testing so I believe this bug got introduced in the whole re-structuring process.

I still cannot confirm it as I don't have the working copy of this version. I will have to test it to verify anything.

**Comment by tuhinjubcse** [comment]
You are the first author of the paper . If you are putting a github link on the camera ready version of the paper I will assume it to be perfectly fine . I would appreciate if you test with the fixed code and put the correct numbers on a Read me or arxiv .

**Comment by shashankg7** [comment]
Hi Tuhin,

Thank you for catching this typo in the code. Actually this got introduced when we were re-factoring our code for the final camera-ready version. As Pinkesh mentioned, we didn't use the args format during the experimentation and added it later. I am copy-pasting the LSTM training code used during experimentation from our previous commit (without args format):

def train_LSTM(X, y, model, inp_dim, weights, epochs=10, batch_size=512):
    cv_object = KFold(n_splits=10, shuffle=True, random_state=42)
    print cv_object
    p, r, f1 = 0., 0., 0.
    p1, r1, f11 = 0., 0., 0.
    sentence_len = X.shape[1]
    for train_index, test_index in cv_object.split(X):
        shuffle_weights(model)
        model.layers[0].set_weights([weights])
	X_train, y_train = X[train_index], y[train_index]
        X_test, y_test = X[test_index], y[test_index]
        #pdb.set_trace()
        y_train = y_train.reshape((len(y_train), 1))
        X_temp = np.hstack((X_train, y_train))
        for epoch in xrange(epochs):
            for X_batch in batch_gen(X_temp, batch_size):
                x = X_batch[:, :sentence_len]
                y_temp = X_batch[:, sentence_len]
		try:
                    y_temp = np_utils.to_categorical(y_temp, nb_classes=3)
                except Exception as e:
                    print e
                    print y_temp
                print x.shape, y.shape
                loss, acc = model.train_on_batch(x, y_temp)#, class_weight=class_weights)
                print loss, acc

        y_pred = model.predict_on_batch(X_test)
        y_pred = np.argmax(y_pred, axis=1)
        print classification_report(y_test, y_pred)
        print precision_recall_fscore_support(y_test, y_pred)
        print y_pred
        p += precision_score(y_test, y_pred, average='weighted')
        p1 += precision_score(y_test, y_pred, average='micro')
        r += recall_score(y_test, y_pred, average='weighted')
        r1 += recall_score(y_test, y_pred, average='micro')
        f1 += f1_score(y_test, y_pred, average='weighted')
        f11 += f1_score(y_test, y_pred, average='micro')

As you can see, we are shuffling (resetting) the weights for each of the CV run. The numbers reported in the paper are infact correct. 

Apologies for the inconvenience, we will try to fix this when time permits. 

Thanks,
Shashank

**Comment by shashankg7** [comment]
Ref: 

https://github.com/pinkeshbadjatiya/twitter-hatespeech/tree/f0379122d93ec97ef22546a7c92113e6b629fc64
